fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java

```diff
@@ -144,6 +144,54 @@ private static void getAllTables(OptExpression root, List<Table> tables) {
         }
     }
 
+    // get all ref table scan descs within and below root
+    // the operator tree must match the rule pattern we define and now we only support SPJG pattern tree rewrite.
+    // so here LogicalScanOperator's children must be LogicalScanOperator or LogicalJoinOperator
+    public static List<TableScanDesc> getTableScanDescs(OptExpression root) {
+        TableScanContext scanContext = new TableScanContext();
+        OptExpressionVisitor joinFinder = new OptExpressionVisitor<Void, TableScanContext>() {
+            @Override
+            public Void visit(OptExpression optExpression, TableScanContext context) {
+                for (OptExpression child : optExpression.getInputs()) {
+                    child.getOp().accept(this, child, context);
+                }
+                return null;
+            }
+
+            @Override
+            public Void visitLogicalTableScan(OptExpression optExpression, TableScanContext context) {
+                LogicalScanOperator scanOperator = (LogicalScanOperator) optExpression.getOp();
+                Table table = scanOperator.getTable();
+                Integer id = scanContext.getTableIdMap().computeIfAbsent(table, t -> 0);
+                TableScanDesc tableScanDesc = new TableScanDesc(table, id, scanOperator, null);
+                context.getTableScanDescs().add(tableScanDesc);
+                scanContext.getTableIdMap().put(table, ++id);
+                return null;
+            }
+
+            @Override
+            public Void visitLogicalJoin(OptExpression optExpression, TableScanContext context) {
+                for (OptExpression child : optExpression.getInputs()) {
+                    if (child.getOp() instanceof LogicalScanOperator) {
```

Comment:
why not visit the LogicalScanOperator recursive

Reply:
LogicalScanOperator does not has any child.